### DISCUSSION TITLE
Error handling: How are errors intended to be programmatically handled?

### DISCUSSION THREAD

**Discussion by zaunermax** [prisma/prisma#3280] [Q&A] [answered]
We tried to adopt prisma 2 in our nestjs REST api and were asking ourselves how errors are intended to be handled.

Especially in create endpoints where one can create a big database entity. So let's say we have an order entity that has relations to locations, customers, truck drivers, etc. We have a prisma call that connects some ids to the order table and if the ID that is used to connect to other tables does not exist an error is thrown. Right now there is no other way than to look at prismas error string from the error message to decide what error to throw. 

Also we noticed, that we cannot write a general catcher function that looks for the string "not exist" for example to then return a 404 error to the requesting client instead of a 500 error with the full prisma error telling the requesting client it was its error by sending a request with an invalid connection id.

We looked in the documentation and searched for issues/discussions but could not find anything on how errors can be programmatically handled, maybe you have a heads up for us :raised_hands: 

LE: the reason I ask this question is, that checking for strings in error messages is really really prone to cause some errors when upgrading prisma, as error messages can change around and be rephrased at some point I guess :thinking:

**Comment by zaunermax** [comment, marked as answer]
Are the error codes documented somewhere? :thinking: For example we noticed if some entity is not found we get the error code `P2016` - would it be feasable to just react to the error like `if (error.code === 'P2016') throw new NotFoundException()`?

**Reply by zaunermax** [reply]
nvm, I just found the [code](https://github.com/prisma/prisma-engines/blob/4dbefe724ae6df3621eb660a5d0e0402e34189bf/libs/user-facing-errors/src/query_engine.rs#L193). Seems to be a "Query interpretation error" in general :eyes:

**Reply by zaunermax** [reply]
some lines further down there is [an error that seems more appropriate](https://github.com/prisma/prisma-engines/blob/4dbefe724ae6df3621eb660a5d0e0402e34189bf/libs/user-facing-errors/src/query_engine.rs#L212) :thinking: When trying to connect to entities that are not found I get the `P2016` error though... Is that a bug?

**Reply by ryands17** [reply]
Hey @zaunermax 👋 
Currently there are no specific errors documented that can be actually used to determine the exact error and also these codes might change in the near future. You would need to rely on the generic errors for now, emitted by Prisma.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
